File: lint-arwaky/archive/infrastructure/naming_variant_provider.py
from __future__ import annotations

import re

from ..taxonomy import SymbolName, ResponseData, SymbolNameList, NamingError
from ..contract import INamingVariantPort


class PythonNamingVariantProvider(INamingVariantPort):
    """Implementation of identifier naming variants for Python."""

    def get_variant_dict(self, name: SymbolName) -> ResponseData | NamingError:
        """Produces common naming variants mapped by their type."""
        name_str = str(name)
        words = re.findall(
            r"[A-Za-z][a-z0-9]*|[A-Z]+(?=[A-Z][a-z0-9]|\b)|[0-9]+", name_str
        )
        words = [w.lower() for w in words]

        if not words:
            return ResponseData(
                value={
                    "snake_case": name_str,
                    "pascal_case": name_str,
                    "camel_case": name_str,
                    "screaming_snake": name_str.upper(),
                }
            )

        snake_case = "_".join(words)
        _first = str(words[0])
        _rest = [str(w).capitalize() for w in words[1:]]
        camel_case = _first + "".join(_rest)
        pascal_case = "".join(str(w).capitalize() for w in words)
        screaming_snake = snake_case.upper()

        res = {
            "snake_case": snake_case,
            "camel_case": camel_case,
            "pascal_case": pascal_case,
            "screaming_snake": screaming_snake,
        }
        return ResponseData(value=res)

    def build_variants(self, name: SymbolName) -> SymbolNameList | NamingError:
        """Produce common naming variants for a given base variable/function name."""
        name_str = str(name)
        variants_data = self.get_variant_dict(name=name)
        if isinstance(variants_data, NamingError):
            return variants_data

        variants_dict = variants_data.value
        sc = str(variants_dict.get("snake_case", name_str))
        ss = str(variants_dict.get("screaming_snake", name_str.upper()))
        cc = str(variants_dict.get("camel_case", name_str))
        pc = str(variants_dict.get("pascal_case", name_str))

        kebab_case = sc.replace("_", "-")
        variants_set = {name_str, sc, ss, cc, pc, kebab_case}
        return SymbolNameList(values=[SymbolName(value=v) for v in variants_set])

File: lint-arwaky/src-rust/infrastructure/naming_variant_provider.rs
/// naming_variant_provider — Python naming variant generator.
use crate::contract::INamingVariantPort;
use crate::taxonomy::{NamingError, ResponseData, SymbolName, SymbolNameList};
use regex::Regex;

pub struct PythonNamingVariantProvider;

impl PythonNamingVariantProvider {
    pub fn new() -> Self {
        Self
    }

    pub fn get_variant_dict(&self, name: SymbolName) -> Result<ResponseData, NamingError> {
        let name_str = &name.value;
        let words: Vec<String> = Regex::new(r"[A-Za-z][a-z0-9]*|[A-Z]+(?=[A-Z][a-z0-9]|\b)|[0-9]+")
            .unwrap().find_iter(name_str).map(|m| m.as_str().to_lowercase()).collect();
        if words.is_empty() {
            return Ok(ResponseData::new(serde_json::json!({"snake_case": name_str, "pascal_case": name_str, "camel_case": name_str, "screaming_snake": name_str.to_uppercase()})));
        }
        let snake_case = words.join("_");
        let first = words[0].clone();
        let rest: String = words[1..].iter().map(|w| { let mut c = w.chars(); match c.next() { Some(ch) => ch.to_uppercase().to_string() + c.as_str(), None => String::new() } }).collect();
        let camel_case = format!("{}{}", first, rest);
        let pascal_case: String = words.iter().map(|w| { let mut c = w.chars(); match c.next() { Some(ch) => ch.to_uppercase().to_string() + c.as_str(), None => String::new() } }).collect();
        let screaming_snake = snake_case.to_uppercase();
        Ok(ResponseData::new(serde_json::json!({"snake_case": snake_case, "camel_case": camel_case, "pascal_case": pascal_case, "screaming_snake": screaming_snake})))
    }

    pub fn build_variants(&self, name: SymbolName) -> Result<SymbolNameList, NamingError> {
        let name_str = &name.value;
        let data = self.get_variant_dict(name)?;
        let map = data.value.as_object().cloned().unwrap_or_default();
        let sc = map.get("snake_case").and_then(|v| v.as_str()).unwrap_or(name_str).to_string();
        let ss = map.get("screaming_snake").and_then(|v| v.as_str()).unwrap_or(&name_str.to_uppercase()).to_string();
        let cc = map.get("camel_case").and_then(|v| v.as_str()).unwrap_or(name_str).to_string();
        let pc = map.get("pascal_case").and_then(|v| v.as_str()).unwrap_or(name_str).to_string();
        let kebab = sc.replace('_', "-");
        let mut variants: Vec<String> = vec![name_str.clone(), sc, ss, cc, pc, kebab];
        variants.sort();
        variants.dedup();
        Ok(SymbolNameList::new(variants.into_iter().map(|v| SymbolName::new(v)).collect()))
    }
}

#[async_trait::async_trait]
impl INamingVariantPort for PythonNamingVariantProvider {
    async fn get_variant_dict(&self, name: &SymbolName) -> Result<ResponseData, NamingError> {
        unimplemented!()
    }
    async fn build_variants(&self, name: &SymbolName) -> Result<Vec<SymbolName>, NamingError> {
        unimplemented!()
    }
}

